// tco-node-api/test/spec_helper.js
var tco = module.exports = require('../lib/twocheckout.js');

tco.config('901248156', "DC034942-83D1-4AD6-9FD8-460E3F876106", true);

exampleCreditCard = {
  number: '4111111111111111',
  exp_month: 12,
  exp_year: 2019,
  cvv: '123'
};

exampleAddress = {
  name: 'Testing Tester',
  address_1: '123 Test St',
  address_2: 'the attic',
  city: 'Columbus',
  state: 'OH',
  country_code: 'US',
  postal_code: '43123'
};

examplePaymentMethod = {
  credit_card: exampleCreditCard,
  address: exampleAddress
};

exampleCustomer = {
  phone: '555-555-5556',
  currency: 'USD',
  lang: 'en',
  email: 'nobody@example.com',
  payment_method: examplePaymentMethod
};

exampleSale = {
  amount: 1.00,
  merchant_order_id: '123',
  customer: exampleCustomer
};

pagenation = {
  page: 2,
  page_size: 2
};

exampleAuthOnlySale = {
  amount: 1.00,
  merchant_order_id: '123',
  auth_only: true,
  customer: exampleCustomer
};

exampleRefund = {
  amount: 1.00,
  currency_type: "vendor",
  comment: "Refund Issued"
};

exampleRecurringSale = {
  amount: 1.00,
  merchant_order_id: '123',
  recurrence: '1 Month',
  customer: exampleCustomer
};
